storage: prevent reuse of time bounded iterators

tbg opened the pull request:
`(*rocksDBBatch).NewTimeboundedIterator` was caching the iterator in a
way that would return the iterator in future calls to
`(*rocksDBBatch).NewIterator`, which was a correctness problem.

Luckily, the fallout was somewhat limited by the fact that we only use
time bounded iterators when resolving an intent, and that is usually
(but not always) the last operation carried out on the batch.

This was introduced in #21078, which landed around Jan 15 2018, so
the bug has not made it into a release yet, though we'll have to bump
our alpha (https://github.com/cockroachdb/cockroach/issues/21657).

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/21721)
<!-- Reviewable:end -->

tbg committed:
storage: prevent reuse of time bounded iterators

`(*rocksDBBatch).NewTimeboundedIterator` was caching the iterator in a
way that would return the iterator in future calls to
`(*rocksDBBatch).NewIterator`, which was a correctness problem.

Luckily, the fallout was somewhat limited by the fact that we only use
time bounded iterators when resolving an intent, and that is usually
(but not always) the last operation carried out on the batch.

This was introduced in #21078, which landed around Jan 15 2018, so
the bug has not made it into a release yet, though we'll have to bump
our alpha (https://github.com/cockroachdb/cockroach/issues/21657).

Release note: None

Fixes #21689.

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

We should also add a test that would have caught this, but let's merge it as-is first so we can get it in a new alpha candidate quickly.

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21721#-:-L3ZnQQ2mJKfar0elhXn:b-vd1gzj)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
@bdarnell Yep, I'm writing the test as we speak, just want to get this out the door.

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21721#-:-L3Zom4kJKmDRiSD4gC1:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
TFTRs. Test PR is here: https://github.com/cockroachdb/cockroach/pull/21723

knz commented:
nit: I'd argue a "bug fix" release note was warranted here.

petermattis commented:
Didn't we determine the bug never made it into a release?

tschottdorf commented:
@knz what @petermattis said. The bug was fixed before leaking out into the world.

tschottdorf commented:
(that said, that's something I should've mentioned more explicitly in the commit).
